File: src/solutions/day01/day01.py
from typing import List
BASE_PATH = "../../inputs/day01"

def get_numbers(file_path: str) -> List[int]:
    with open(file_path) as input_file:
        return [int(l.strip()) for l in input_file if l.strip()]

def generic_solution(numbers: List[int], window_size=1) -> int:
    return sum([int(numbers[i] < numbers[i+window_size]) for i in range(len(numbers)-window_size)])

def solution_1(numbers: List[int]) -> int:
    return generic_solution(numbers, window_size=1)

def solution_2(numbers: List[int]) -> int:
    return generic_solution(numbers, window_size=3)



for file_name in ["test.txt", "sol.txt"]:
    print(f"Solutions for {file_name}:")
    numbers = get_numbers(f"../../inputs/day01/{file_name}")
    print(f"\tPart 1: {solution_1(numbers)}")
    print(f"\tPart 2: {solution_2(numbers)}")

File: src/solutions/day01/day01.rs
fn get_numbers(file_content: &str) -> Vec<usize> {
    file_content
        .lines()
        .map(|line| line.parse::<usize>().expect("Can't parse number"))
        .collect::<Vec<usize>>()
}

fn generic_solution(numbers: &Vec<usize>, window_size: usize) -> usize {
    numbers
        .windows(window_size)
        .collect::<Vec<&[usize]>>()
        .windows(2)
        .into_iter()
        .filter(|windows| windows[0][0] < windows[1][windows[1].len() - 1])
        .count()
}

pub fn day01() {
    println!("Solutions for test.txt:");
    let numbers = get_numbers(include_str!("../../inputs/day01/test.txt"));
    println!("Part 1 {}", generic_solution(&numbers, 1));
    println!("Part 2 {}", generic_solution(&numbers, 3));

    println!("Solutions for day01.txt:");
    let numbers = get_numbers(include_str!("../../inputs/day01/sol.txt"));
    println!("Part 1 {}", generic_solution(&numbers, 1));
    println!("Part 2 {}", generic_solution(&numbers, 3));
}
